### PR TITLE
[contractkit] SortedOracles: check for and remove expired reports

### PR DESCRIPTION
**Pull request by yerdua**
## Description

In order to facilitate cleaning up after oracles that have stopped reporting, this PR adds two functions to the SortedOracles wrapper: `removeExpiredReports` and `isOldestReportExpired`.

## Other changes

Fixed a misleading typo in a function comment

## Tested

This PR includes tests for the new functions.

## Backwards compatibility

Yes! Functions are added to the wrapper, not modified or removed.

## Commit Message
[contractkit] SortedOracles: check for and remove expired reports

In order to facilitate cleaning up after oracles that have stopped reporting, two functions are added to the SortedOracles wrapper: `removeExpiredReports` and `isOldestReportExpired`.

### CODE REVIEW

File: packages/contractkit/src/wrappers/SortedOracles.ts [hunk 1]
```diff
@@ -105,9 +105,35 @@ export class SortedOraclesWrapper extends BaseWrapper<SortedOracles> {
...
   )
 
+  /**
+   * Checks if the oldest report for a given token is expired
+   * @param token The token for which to check reports
+   */
+  async isOldestReportExpired(token: CeloToken): Promise<[boolean, Address]> {
+    const tokenAddress = await this.kit.registry.addressFor(token)
+    const response = await this.contract.methods.isOldestReportExpired(tokenAddress).call()
+    return response as [boolean, Address]
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.ts:115]: +1
this is much nicer than that I have done in the past with other multiple value returns
another thing that web3 types does poorly...

File: packages/contractkit/src/wrappers/SortedOracles.ts [hunk 2]
```diff
@@ -105,9 +105,35 @@ export class SortedOraclesWrapper extends BaseWrapper<SortedOracles> {
...
+  }
+
+  /**
+   * Removes expired reports, if any exist
+   * @param token The token to remove reports for
+   * @param numReports The number of expired reports to remove.
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.ts]: can this be adapted to explain upper bound nature (ie numReports=5 can remove only 1 report and still be a successful transaction)

File: packages/contractkit/src/wrappers/SortedOracles.test.ts [hunk 3]
```diff
@@ -115,6 +122,81 @@ testWithGanache('SortedOracles Wrapper', (web3) => {
     })
   })
 
+  describe('#removeExpiredReports', () => {
+    beforeEach(async () => {
+      await reportAsOracles(stableTokenOracles)
+    })
+
+    describe('when expired reports exist', () => {
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.test.ts:136]: nit: `describe`+`beforeEach` distilled to single `it` block

File: packages/contractkit/src/wrappers/SortedOracles.test.ts [hunk 4]
```diff
@@ -115,6 +122,81 @@ testWithGanache('SortedOracles Wrapper', (web3) => {
...
+        const initialReportCount = await sortedOracles.numRates(CeloContract.StableToken)
+        const tx = await sortedOracles.removeExpiredReports(CeloContract.StableToken, 1)
+        await tx.sendAndWaitForReceipt({ from: oracleAddress })
+        expect(await sortedOracles.numRates(CeloContract.StableToken)).toEqual(
+          initialReportCount - 1
+        )
+      })
+    })
+
+    describe('when reports exist but are not expired', () => {
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.test.ts]: ditto

File: packages/contractkit/src/wrappers/SortedOracles.test.ts [hunk 5]
```diff
@@ -115,6 +122,81 @@ testWithGanache('SortedOracles Wrapper', (web3) => {
...
+  })
+
+  describe('#isOldestReportExpired', () => {
+    let expirySeconds: number
+
+    beforeEach(async () => {
+      expirySeconds = (await sortedOracles.reportExpirySeconds()).toNumber()
+    })
+
+    describe('when at least one expired report exists', () => {
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.test.ts]: ditto?
style choice at this point

File: packages/contractkit/src/wrappers/SortedOracles.test.ts [hunk 6]
```diff
@@ -115,6 +122,81 @@ testWithGanache('SortedOracles Wrapper', (web3) => {
...
+
+      it('returns with true and the address of the last reporting oracle', async () => {
+        const [isExpired, address] = await sortedOracles.isOldestReportExpired(
+          CeloContract.StableToken
+        )
+        expect(isExpired).toEqual(true)
+        expect(address).toEqual(oracleAddress)
+      })
+    })
+    describe('when the oldest is not expired', () => {
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.test.ts:188]: ditto?
style choice at this point
consider de-duping some of this

File: packages/contractkit/src/wrappers/SortedOracles.test.ts [hunk 7]
```diff
@@ -125,31 +207,21 @@ testWithGanache('SortedOracles Wrapper', (web3) => {
    * thing to check is if there have been changes to the migrations
    */
   describe('#getRates', () => {
+    const expectedRates = [2, 1.5, 1, 0.5]
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.test.ts:210]: +1
much more readable

File: packages/contractkit/src/wrappers/SortedOracles.ts [hunk 8]
```diff
@@ -105,9 +105,35 @@ export class SortedOraclesWrapper extends BaseWrapper<SortedOracles> {
...
+    return toTransactionObject(
+      this.kit,
+      this.contract.methods.removeExpiredReports(tokenAddress, numReports)
+    )
+  }
+
   /**
    * Updates an oracle value and the median.
-   * @param token The address of the token for which the Celo Gold exchange rate is being reported.
+   * @param token The token for which the Celo Gold exchange rate is being reported.
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/SortedOracles.ts:138]: nice catch

**Review by yorhodes — APPROVED** [review]
just some nits in the test and docstrings

File: packages/contractkit/src/wrappers/SortedOracles.ts (hunk 2 above)
Comment (yerdua) [reply packages/contractkit/src/wrappers/SortedOracles.ts]: Yes! Thanks for the clarity-suggestion :)

File: packages/contractkit/src/wrappers/SortedOracles.test.ts (hunk 5 above)
Comment (yerdua) [reply packages/contractkit/src/wrappers/SortedOracles.test.ts]: I think in some of these cases it aids readability - separating the setup with the expectation-checks. The ones I'd like to leave as-is are the ones where the setup is more than a couple lines, and/or the combined text of `describe` + `it` would extend past the line width limit.

**Pull request merged**